### PR TITLE
Add SQLite support, fix un-thread-safe DB caches, small performance f…

### PR DESCRIPTION
**Pull request by NyaaaWhatsUpDoc**
- adds sqlite support as a DB type, handling warnings for in-memory DBs if selected
- small performance fixes in the DB code (just reducing no. loop iterations)
- add StatusCache implementation that supports key lookup by ID,URI,URL
- fix current DB cache implementations to be thread-safe

### CODE REVIEW

File: internal/db/bundb/session.go [hunk 1]
```diff
@@ -23,22 +23,19 @@ import (
...
 type sessionDB struct {
 	config *config.Config
-	conn   *bun.DB
-	log    *logrus.Logger
+	conn   *DBConn
 }
 
 func (s *sessionDB) GetSession(ctx context.Context) (*gtsmodel.RouterSession, db.Error) {
-	rss := []*gtsmodel.RouterSession{}
+	rss := make([]*gtsmodel.RouterSession, 0, 1)
```
Comment (tsmethurst) [inline internal/db/bundb/session.go:38]: nice

File: internal/db/bundb/status.go [hunk 2]
```diff
@@ -84,15 +53,19 @@ func (s *statusDB) newStatusQ(status interface{}) *bun.SelectQuery {
 
 func (s *statusDB) getAttachedStatuses(ctx context.Context, status *gtsmodel.Status) *gtsmodel.Status {
 	if status.InReplyToID != "" && status.InReplyTo == nil {
-		if inReplyTo, cached := s.statusCached(status.InReplyToID); cached {
+		// TODO: do we want to keep this possibly recursive strategy?
```
Comment (tsmethurst) [inline internal/db/bundb/status.go:56]: probably not, expensive

File: internal/db/bundb/status.go [hunk 3]
```diff
@@ -84,15 +53,19 @@ func (s *statusDB) newStatusQ(status interface{}) *bun.SelectQuery {
...
+		if inReplyTo, cached := s.cache.GetByID(status.InReplyToID); cached {
 			status.InReplyTo = inReplyTo
 		} else if inReplyTo, err := s.GetStatusByID(ctx, status.InReplyToID); err == nil {
 			status.InReplyTo = inReplyTo
 		}
 	}
 
 	if status.BoostOfID != "" && status.BoostOf == nil {
-		if boostOf, cached := s.statusCached(status.BoostOfID); cached {
+		// TODO: do we want to keep this possibly recursive strategy?
```
Comment (tsmethurst) [inline internal/db/bundb/status.go:66]: same comment

File: internal/db/bundb/status_test.go [hunk 4]
```diff
@@ -59,7 +59,6 @@ func (suite *StatusTestSuite) TearDownTest() {
 func (suite *StatusTestSuite) TestGetStatusByID() {
 	status, err := suite.db.GetStatusByID(context.Background(), suite.testStatuses["local_account_1_status_1"].ID)
 	if err != nil {
-		fmt.Println(err.Error())
```
Comment (tsmethurst) [inline internal/db/bundb/status_test.go:62]: woops

File: internal/db/bundb/timeline.go [hunk 5]
```diff
@@ -32,12 +31,18 @@ import (
...
 type timelineDB struct {
 	config *config.Config
-	conn   *bun.DB
-	log    *logrus.Logger
+	conn   *DBConn
 }
 
 func (t *timelineDB) GetHomeTimeline(ctx context.Context, accountID string, maxID string, sinceID string, minID string, limit int, local bool) ([]*gtsmodel.Status, db.Error) {
-	statuses := []*gtsmodel.Status{}
+	// Ensure sane value
```
Comment (tsmethurst) [inline internal/db/bundb/timeline.go:38]: *sensible/reasonable

File: internal/db/bundb/timeline.go [hunk 6]
```diff
@@ -86,11 +91,21 @@ func (t *timelineDB) GetHomeTimeline(ctx context.Context, accountID string, maxI
...
+	err := q.Scan(ctx)
+	if err != nil {
+		return nil, t.conn.ProcessError(err)
+	}
+	return statuses, nil
 }
 
 func (t *timelineDB) GetPublicTimeline(ctx context.Context, accountID string, maxID string, sinceID string, minID string, limit int, local bool) ([]*gtsmodel.Status, db.Error) {
-	statuses := []*gtsmodel.Status{}
+	// Ensure sane value
```
Comment (tsmethurst) [inline internal/db/bundb/timeline.go:102]: *sensible/reasonable

File: internal/db/bundb/timeline.go [hunk 7]
```diff
@@ -121,14 +136,23 @@ func (t *timelineDB) GetPublicTimeline(ctx context.Context, accountID string, ma
...
+	if err != nil {
+		return nil, t.conn.ProcessError(err)
+	}
+	return statuses, nil
 }
 
 // TODO optimize this query and the logic here, because it's slow as balls -- it takes like a literal second to return with a limit of 20!
 // It might be worth serving it through a timeline instead of raw DB queries, like we do for Home feeds.
 func (t *timelineDB) GetFavedTimeline(ctx context.Context, accountID string, maxID string, minID string, limit int) ([]*gtsmodel.Status, string, string, db.Error) {
+	// Ensure sane value
```
Comment (tsmethurst) [inline internal/db/bundb/timeline.go:149]: *sensible/reasonable

**Review by tsmethurst — COMMENTED** [review]
Just a few very small changes

File: internal/db/bundb/status.go (hunk 2 above)
Comment (NyaaaWhatsUpDoc) [reply internal/db/bundb/status.go]: I'll leave it as a TODO for now, we can handle it with the other DB changes

**Review by tsmethurst — APPROVED** [review]

**Pull request merged**